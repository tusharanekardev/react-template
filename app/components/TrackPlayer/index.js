/**
 *
 * TrackPlayer
 *
 */

import React, { useEffect } from 'react';
// import PropTypes from 'prop-types'
// import styled from 'styled-components'
import styled from '@emotion/styled';
import PropTypes from 'prop-types';

const TrackPlayerDiv = styled.div`
  position: fixed;
  bottom: 0;
  width: 600px;
  background-color: #d3d3d3;
  display: flex;
  padding: 0.5rem;
  border-radius: 10px;
  gap: 1rem;
  height: 80px;
  left: 50%;
  transform: translateX(-50%);
`;

/**
 * TrackPlayer component that renders the current track's information
 * and provides play/pause functionality.
 *
 * @component
 *
 * @param {Object} props - The props object.
 * @param {number|string} props.playTrackId - The ID of the track currently selected for playback.
 * @param {Object} props.tracksData - An object containing track data, including an array of track results.
 * @param {boolean} props.isPlaying - A boolean indicating whether a track is currently playing.
 * @param {function} props.dispatchIsPlaying - A function to dispatch the action for toggling the play/pause state.
 *
 * @description
 * - Displays information about the current track (such as the track name, artist name, and artwork).
 * - Provides a button to toggle between playing and pausing the track.
 * - Plays the track automatically when a new `playTrackId` is provided.
 *
 * @returns {JSX.Element} The rendered TrackPlayer component.
 */
export function TrackPlayer({ playTrackId, tracksData, isPlaying, dispatchIsPlaying }) {
  /**
   * Finds the current track being played from the list of tracks.
   *
   * @constant {Object|undefined} currentTrack
   * @description
   * - Searches for the track with a `trackId` that matches the `playTrackId` in the `tracksData.results` array.
   * - If `tracksData` or `results` is undefined, it safely returns `undefined` without throwing an error.
   *
   * @param {Array<Object>} [tracksData.results] - The array of track objects from which to find the current track.
   * @param {number|string} playTrackId - The ID of the track currently being played.
   * @returns {Object|undefined} The track object with the matching `trackId`, or `undefined` if no match is found.
   */
  const currentTrack = tracksData?.results?.find((track) => track?.trackId === playTrackId);
  const { artistName = '', previewUrl = '', artworkUrl100 = '', trackId = '', trackName = '' } = currentTrack || {};
  let audio = document.getElementById('audio_tag');

  /**
   * Toggles the play/pause state of a track.
   *
   * @function togglePlayPause
   * @param {string | number} trackId - The ID of the track to be played or paused.
   * @description
   * - Sets the currently playing track using `setPlayTrackId`.
   * - If the track is not already playing, dispatches an action to start playing by calling `dispatchIsPlaying`.
   */
  function handleTogglePlayPause() {
    if (isPlaying) {
      audio.pause();
    } else {
      audio.play();
    }
    dispatchIsPlaying();
  }

  useEffect(() => {
    audio = document.getElementById('audio_tag');
    if (audio) {
      audio.play();
    }
  }, [playTrackId]);

  return (
    <TrackPlayerDiv data-testid="track-player">
      {trackId ? (
        <>
          <img src={artworkUrl100} alt={`${trackName} cover`} />
          <div>
            <p>{trackName}</p>
            <p>{artistName}</p>
          </div>
          <button onClick={handleTogglePlayPause}>{isPlaying ? 'Pause' : 'Play'}</button>
          <audio id="audio_tag" src={previewUrl} />
        </>
      ) : (
        <p>Select song to play</p>
      )}
    </TrackPlayerDiv>
  );
}

TrackPlayer.propTypes = {
  playTrackId: PropTypes.number,
  tracksData: PropTypes.shape({
    resultCount: PropTypes.number,
    results: PropTypes.array
  }),
  isPlaying: PropTypes.bool,
  dispatchIsPlaying: PropTypes.func
};

export default TrackPlayer;
