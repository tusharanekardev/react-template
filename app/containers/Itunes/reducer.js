/*
 *
 * Itunes reducer
 *
 */
import { produce } from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const initialState = {
  trackName: null,
  tracksData: {},
  tracksError: null,
  loading: null,
  trackId: null,
  trackData: {},
  trackError: null,
  isPlaying: false
};

export const { Types: itunesTypes, Creators: itunesCreators } = createActions({
  requestGetItunesTracks: ['trackName'],
  successGetItunesTracks: ['data'],
  failureGetItunesTracks: ['error'],
  requestGetItunesTrackById: ['trackId'],
  successGetItunesTrackById: ['trackData'],
  failureGetItunesTrackById: ['trackError'],
  toggleIsPlaying: ['isPlaying']
});

/* eslint-disable complexity */
// Code with higher complexity
export const itunesReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case itunesTypes.REQUEST_GET_ITUNES_TRACKS:
        draft.trackName = action.trackName;
        draft.loading = true;
        break;

      case itunesTypes.SUCCESS_GET_ITUNES_TRACKS:
        draft.tracksData = action.data;
        draft.tracksError = null;
        draft.loading = false;
        break;

      case itunesTypes.FAILURE_GET_ITUNES_TRACKS:
        draft.tracksError = get(action.error, 'message', 'something went wrong');
        draft.tracksData = null;
        draft.loading = false;
        break;

      case itunesTypes.REQUEST_GET_ITUNES_TRACK_BY_ID:
        draft.trackId = action.trackId;
        draft.loading = true;
        break;

      case itunesTypes.SUCCESS_GET_ITUNES_TRACK_BY_ID:
        draft.trackData = action.data;
        draft.trackError = null;
        draft.loading = false;
        break;

      case itunesTypes.FAILURE_GET_ITUNES_TRACK_BY_ID:
        draft.trackError = get(action.error, 'message', 'something went wrong');
        draft.trackData = null;
        draft.loading = false;
        break;

      case itunesTypes.TOGGLE_IS_PLAYING:
        draft.isPlaying = !state.isPlaying;
        break;
    }
  });

export default itunesReducer;
