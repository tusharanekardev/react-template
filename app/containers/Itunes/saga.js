import { put, call, takeLatest } from 'redux-saga/effects';
import { itunesCreators, itunesTypes } from './reducer';
import { getTracks } from '@app/services/trackApi';

const { REQUEST_GET_ITUNES_TRACKS } = itunesTypes;
const { successGetItunesTracks, failureGetItunesTracks } = itunesCreators;

// Individual exports for testing
/**
 * A saga that handles fetching GitHub repositories based on a given repository name.
 * On success, it dispatches a success action with the fetched data.
 * On failure, it dispatches a failure action with the error data.
 *
 * @date 01/03/2024 - 14:47:28
 *
 * @param {Object} action - The action object containing the repository name.
 * @yields {Effect} The effect of calling the API, and then either the success or failure action.
 */
export function* getItunesTracks(action) {
  const response = yield call(getTracks, action.trackName);
  const { data, ok } = response;
  if (ok) {
    yield put(successGetItunesTracks(data));
  } else {
    yield put(failureGetItunesTracks(data));
  }
}

/**
 * registering events
 * @date 04/03/2024 - 12:57:49
 *
 * @export
 * @returns {{}}
 */
export default function* itunesSaga() {
  yield takeLatest(REQUEST_GET_ITUNES_TRACKS, getItunesTracks);
}
