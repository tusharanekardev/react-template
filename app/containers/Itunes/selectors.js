import { createSelector } from 'reselect';
import get from 'lodash/get';
import { initialState } from './reducer';

/**
 * Direct selector to the itunes state domain
 */

export const selectItunesDomain = (state) => state.itunesContainer || initialState;

export const selectItunesData = () => createSelector(selectItunesDomain, (substate) => get(substate, 'tracksData'));

export const selectLoading = () => createSelector(selectItunesDomain, (substate) => get(substate, 'loading'));

export const selectTracksError = () => createSelector(selectItunesDomain, (substate) => get(substate, 'tracksError'));

export const selectTrackName = () => createSelector(selectItunesDomain, (substate) => get(substate, 'trackName'));

export const selectTrackId = () => createSelector(selectItunesDomain, (substate) => get(substate, 'trackId'));

export const selectTrackData = () => createSelector(selectItunesDomain, (substate) => get(substate, 'trackData'));

export const selectTrackError = () => createSelector(selectItunesDomain, (substate) => get(substate, 'trackError'));
