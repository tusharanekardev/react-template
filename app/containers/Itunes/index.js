import React, { useEffect, memo } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import debounce from 'lodash/debounce';
import isEmpty from 'lodash/isEmpty';
import { createStructuredSelector } from 'reselect';
import { For } from '@components/For';
import { compose } from 'redux';
import styled from '@emotion/styled';
import { injectSaga } from 'redux-injectors';
import { OutlinedInput } from '@mui/material';
import { selectItunesData, selectLoading, selectTrackName } from './selectors';
import itunesSaga from './saga';
import { itunesCreators } from './reducer';
import { translate } from '@app/utils';
import { TrackCard } from '@components/TrackCard';

const Container = styled.div`
  && {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr;
    gap: 1rem;
    margin-top: 1rem;
  }
`;
const StyledOutlinedInput = styled(OutlinedInput)`
  legend {
    display: none;
  }
  > fieldset {
    top: 0;
  }
`;

/**
 * Itunes component that handles searching and displaying iTunes tracks.
 * It includes input handling for track searches, loading state management,
 * and rendering of the track list using the TrackCard component.
 *
 * The component fetches track data based on user input and displays the results
 * in a grid layout. It also debounces input changes to optimize API calls.
 *
 * @component
 * @date 01/03/2024 - 14:47:28
 *
 * @param {Object} props - The component props.
 * @param {Function} props.dispatchItunesTracks - Function to dispatch the action for fetching iTunes tracks based on the provided track name.
 * @param {boolean} props.loading - Indicates whether the tracks are currently being loaded.
 * @param {string} props.trackName - The current track name input by the user for searching tracks.
 * @param {Object} props.tracksData - The data object containing the results of the fetched iTunes tracks.
 * @param {number} props.tracksData.resultCount - The total number of results fetched from the iTunes API.
 * @param {Array} props.tracksData.results - The array of track objects fetched from the iTunes API.
 *
 * @returns {JSX.Element} The rendered Itunes component.
 */
export function Itunes({ dispatchItunesTracks, loading, trackName, tracksData }) {
  useEffect(() => {
    if (trackName && !tracksData?.results?.length) {
      dispatchItunesTracks(trackName);
    }
  }, []);

  const searchTracks = (tName) => {
    dispatchItunesTracks(tName);
  };

  const handleOnChange = (tName) => {
    if (!isEmpty(tName)) {
      searchTracks(tName);
    } else {
      // dispatchClearGithubRepos();
    }
  };

  const debouncedHandleOnChange = debounce(handleOnChange, 200);

  return (
    <div>
      <StyledOutlinedInput
        inputProps={{ 'data-testid': 'search-bar' }}
        onChange={(event) => debouncedHandleOnChange(event.target.value)}
        fullWidth
        defaultValue={trackName}
        placeholder={translate('default_template')}
      />

      <For
        of={tracksData?.results}
        ParentComponent={Container}
        renderItem={(item, index) => <TrackCard key={index} {...item} />}
      />
    </div>
  );
}

Itunes.propTypes = {
  dispatchItunesTracks: PropTypes.func,
  loading: PropTypes.bool,
  trackName: PropTypes.string,
  tracksData: PropTypes.shape({
    resultCount: PropTypes.number,
    results: PropTypes.array
  })
};

const mapStateToProps = createStructuredSelector({
  loading: selectLoading(),
  tracksData: selectItunesData(),
  trackName: selectTrackName()
});

/**
 * Maps dispatch actions to props for the Itunes component.
 * This function binds the action creator for fetching iTunes tracks
 * to the component's props, allowing the component to dispatch
 * actions to the Redux store.
 *
 * @function
 * @param {Function} dispatch - The Redux dispatch function.
 * @returns {Object} An object containing the dispatch functions to be used in the component.
 * @returns {Function} return.dispatchItunesTracks - Function to dispatch the requestGetItunesTracks action
 *    with the specified track name.
 */
export function mapDispatchToProps(dispatch) {
  const { requestGetItunesTracks } = itunesCreators;
  return {
    dispatchItunesTracks: (trackName) => dispatch(requestGetItunesTracks(trackName))
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, memo, injectSaga({ key: 'itunes', saga: itunesSaga }))(Itunes);

export const ItunesTest = compose()(Itunes);
