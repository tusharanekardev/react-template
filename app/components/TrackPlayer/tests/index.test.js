/**
 *
 * Tests for TrackPlayer
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { renderProvider, renderWithIntl } from '@utils/testUtils';
import TrackPlayer from '../index';
import { itunesTypes } from '@app/containers/Itunes/reducer';
import { mapDispatchToProps } from '@app/containers/Itunes/index';

describe('<TrackPlayer />', () => {

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<TrackPlayer />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackPlayer component', () => {
    const { getAllByTestId } = renderProvider(<TrackPlayer />);
    expect(getAllByTestId('track-player').length).toBe(1);
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchTracksSerachSpy = jest.fn()
    const trackName = 'Beatles'
    const actions = {
      dispatchIsPlaying: {type: itunesTypes.TOGGLE_IS_PLAYING}
    }

    const props = mapDispatchToProps(dispatchTracksSerachSpy)
    props.dispatchIsPlaying()
    expect(dispatchTracksSerachSpy).toHaveBeenCalledWith(actions.dispatchIsPlaying)

    props.dispatchIsPlaying()
  })
});
