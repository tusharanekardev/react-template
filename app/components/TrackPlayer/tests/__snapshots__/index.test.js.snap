// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackPlayer /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="css-1vmggjl e1hn6k2i0"
      data-testid="track-player"
    >
      <p>
        Select song to play
      </p>
    </div>
  </div>
</body>
`;
